Move remaining cert helper functions from pkg/serviceaccount to client-go/util/cert

cblecker opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Unifies all remaining certificate helper functions from pkh/serviceaccount to client-go/util/cert. Private key functions were moved in #40907

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #40807

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

cblecker commented:
cc: @liggitt regarding #40807

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41611)
<!-- Reviewable:end -->

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -55,6 +56,33 @@ func EncodeCertPEM(cert *x509.Certificate) []byte {
 }
 
+// ReadPrivateKeyFromFile is a helper function for reading a private key from a PEM-encoded file
+func ReadPrivateKeyFromFile(file string) (interface{}, error) {
```
Belongs in io.go

cblecker replied:
Fixed.

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -55,6 +56,33 @@ func EncodeCertPEM(cert *x509.Certificate) []byte {
+
+// ReadPublicKeysFromFile is a helper function for reading x509.Certificates from a PEM-encoded file.
+// Reads public keys from both public and private key files.
+func ReadPublicKeysFromFile(file string) ([]interface{}, error) {
```
io.go

cblecker replied:
Fixed.

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -117,7 +176,7 @@ func ParseCertsPEM(pemCerts []byte) ([]*x509.Certificate, error) {
 
 	if !ok {
-		return certs, errors.New("could not read any certificates")
+		return certs, errors.New("data does not contain a valid RSA or ECDSA certificates")
```
This message doesn't make sense

cblecker replied:
Fixed.

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -91,6 +119,37 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
 }
 
+// ParsePublicKeysPEM is a helper function for reading an array of rsa.PublicKey or ecdsa.PublicKey from a PEM-encoded byte array.
+// Reads public keys from both public and private key files.
```
The impl doesn't match the godoc, and dropped support for reading public keys from private key blocks.

liggitt replied:
Was there a good reason not to use the existing impl as-is?

cblecker replied:
The comment was retained from the function it replaces here: https://github.com/kubernetes/kubernetes/pull/41611/files#diff-cad4f907f9a840865e4ccf6d2bdf7e7bL90

However, looking over the code, as written does not actually parse out public keys from private key files.. it simply discards them as they wouldn't match "PUBLIC KEY" in the pem headers.

Fixing functionality to match intended function

liggitt replied:
the existing code did parse them... it didn't check block headers at all

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -91,6 +92,46 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
+				return keys, err
+			}
+			keys = append(keys, publickey)
+		case "PRIVATE KEY":
```
This doesn't handle explicit RSA or ECDSA private key blocks

cblecker replied:
Thanks for your patience in reviewing. Fixed!

cblecker commented:
Hi @liggitt! Any feedback on the most recent changes? Thanks again for your time.

jcbsmpsn commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -91,6 +91,63 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
+			break
+		}
+		// Only use PEM "PUBLIC KEY" or "PRIVATE KEY" blocks without extra headers
+		if block.Type != "PUBLIC KEY" || block.Type != "PRIVATE KEY" || len(block.Headers) != 0 {
```
If you rebase this PR, there are some constants defined for these strings at the top of this file.

cblecker replied:
Rebased, and modified. Thanks!

liggitt commented:
push past past code freeze to keep the queue available for things required for 1.6. will take a look on Monday

liggitt commented on pkg/serviceaccount/jwt_test.go:
```diff
@@ -107,12 +107,12 @@ X2i8uIp/C/ASqiIGUeeKQtX0/IR3qCXyThP/dbCiHrF3v1cuhBOHY8CLVg==
+	keys, _ := certutil.ParsePublicKeysPEM([]byte(data))
 	return keys[0]
 }
 func TestReadPrivateKey(t *testing.T) {
```
can move the tests for the relocated functions to the certutil package as well

cblecker replied:
fixed

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -106,6 +106,63 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
+				return keys, err
+			}
+			var parsedKey *rsa.PrivateKey
+			parsedKey, ok = privateKey.(*rsa.PrivateKey)
```
can this be an ECDSA key as well?

cblecker replied:
I believe it can, although I don't think we're using it. either way, test case has been added.

cblecker commented:
Sorry, forgot to actually comment.. I pushed a new commit a few days ago, @liggitt. Let me know what you think when you get a chance. Thank you!

liggitt commented on staging/src/k8s.io/client-go/util/cert/io.go:
```diff
@@ -135,16 +135,30 @@ func CertsFromFile(file string) ([]*x509.Certificate, error) {
 // Returns an error if the file could not be read or if the private key could not be parsed.
-func PrivateKeyFromFile(file string) (interface{}, error) {
-	pemBlock, err := ioutil.ReadFile(file)
+func ReadPrivateKeyFromFile(file string) (interface{}, error) {
```
don't change the name of this function, we treat client-go as an exported API, and the new name isn't tremendously better

cblecker replied:
Ah! Yeah, that makes total sense. Will fix.

liggitt commented on staging/src/k8s.io/client-go/util/cert/io.go:
```diff
@@ -135,16 +135,30 @@ func CertsFromFile(file string) ([]*x509.Certificate, error) {
+
+// ReadPublicKeysFromFile is a helper function for reading x509.Certificates from a PEM-encoded file.
+// Reads public keys from both public and private key files.
+func ReadPublicKeysFromFile(file string) ([]interface{}, error) {
```
follow the naming pattern in the file - `PublicKeysFromFile`

cblecker replied:
fixed

liggitt commented on staging/src/k8s.io/client-go/util/cert/io.go:
```diff
@@ -135,16 +135,30 @@ func CertsFromFile(file string) ([]*x509.Certificate, error) {
 	return key, nil
 }
+
+// ReadPublicKeysFromFile is a helper function for reading x509.Certificates from a PEM-encoded file.
```
fix godoc, this doesn't read x509.Certificates

cblecker replied:
fixed

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -106,6 +107,64 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
+			break
+		}
+		// Only use PEM "PUBLIC KEY" or "PRIVATE KEY" blocks without extra headers
+		if block.Type != PublicKeyBlockType || block.Type != PrivateKeyBlockType || len(block.Headers) != 0 {
```
this check guarantees we'll never reach the ECPrivateKeyBlockType and RSAPrivateKeyBlockType cases, and combining with `||` guarantees we'll always `continue`

liggitt replied:
not actually sure why we need to do this at all... we weren't checking block.Headers before. I would just switch on the block type

cblecker replied:
yeah, this looks to be unneeded, as we're just doing a switch based off what the block is. if it doesn't match, then we'll skip the block and on to the next

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -106,6 +107,64 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
+				keys = append(keys, &parsedECKey.PublicKey)
+			} else {
+				return keys, err
+			}
```
`RSA PUBLIC KEY` and `ECDSA PUBLIC KEY` are also PEM blocks we should accept

liggitt replied:
I don't mind checking block types, but we definitely need unit tests to make sure all the expected formats work

cblecker replied:
Will look into these block types, as well as adding unit tests for them.

cblecker replied:
So I did a bunch of looking into this, and while `RSA PUBLIC KEY` and `EC PUBLIC KEY` are valid PEM block types, it doesn't look like golang's https://golang.org/pkg/crypto/x509/#ParsePKIXPublicKey is set up to handle them. I generated an RSA PUBLIC KEY in that older format, and ParsePKIXPublicKey rejected it flat out. As ParsePKIXPublicKey was what jwt is using, we aren't losing functionality I don't think.

liggitt replied:
ParsePKIXPublicKey doesn't look at PEM block headers at all (and neither did the old code)... it just attempted to decode RSA public keys from the data.

we currently use `RSA PUBLIC KEY` keys successfully as service account keys.

cblecker replied:
I don't see any tests in the code that specifically call a `BEGIN RSA PUBLIC KEY` block, even in the jwt code that serviceaccount uses (https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/dgrijalva/jwt-go/rsa_utils.go#L42-L69). If I pass a `BEGIN RSA PUBLIC KEY` using the old PKCS1 format for public keys, ParsePKIXPublicKey errors out on the block. The only extra thing that I found was that it will also try to parse public keys out of certificates, so I just added that.

Is there somewhere I'm missing that the code calls a `BEGIN RSA PUBLIC KEY` block instead of a `BEGIN PUBLIC KEY` block?

liggitt replied:
these parse key files from users, who have all sorts of key files, some of which even include PKIX rsa key data in a `BEGIN RSA PUBLIC KEY` block :-/

the old code tolerated that sort of insanity by just trying to decode the key types it recognizes. the stdlib tls private key parsing does something similar (https://github.com/golang/go/blob/master/src/crypto/tls/tls.go#L234 and https://github.com/golang/go/blob/master/src/crypto/tls/tls.go#L280-L297)

liggitt commented on vendor/BUILD:
```diff
@@ -16312,3 +16312,9 @@ filegroup(
 )
+
+go_test(
+    name = "k8s.io/client-go/util/cert_xtest",
```
shouldn't this be pem_test.go?

liggitt commented:
were you able to run unit tests locally on this? there were issues I would have expected to fail them

cblecker commented:
@liggitt: what about something like https://github.com/kubernetes/kubernetes/pull/41611/commits/2a08464b2c4704cc6a41f5f3deaacac3f4041bff

cblecker commented:
@k8s-bot test this

cblecker commented:
@k8s-bot gci gce e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot non-cri e2e test this

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -106,6 +108,73 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
+			break
+		}
+
+		switch {
```
I'm still concerned about breaking currently working clusters by strictly enforcing PEM headers (specifically ones mixing up `RSA PRIVATE KEY`/`PRIVATE KEY` and `RSA PUBLIC KEY`/`PUBLIC KEY` PEM headers).

Is there a good reason not to simply try to parse the forms we recognize from the PEM blocks, as we do today?

cblecker commented:
@liggitt The primary reason for some new logic was to prevent requiring the import of `jwt "github.com/dgrijalva/jwt-go"`. However at this point it's probably more complicated than it's worth, and you raise a good point around the risks associated with making the logic more restrictive.

As such, I've brought the logic over with minimal modifications, and brought over the import too.

cblecker commented:
@k8s-bot non-cri e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot gci gce e2e test this

cblecker commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

cblecker commented:
@liggitt The test contexts that are reporting as failed by the bot have been renamed. The only test has has actually failed here is the federation test that is non-blocking and has been flakey.

This should be ready for review.

cblecker commented:
@liggitt Is there time to get this reviewed before code freeze at EOD 6/1? Or should the milestone be bumped?

cblecker commented:
@liggitt Rebased and tests are passing. PTAL

cblecker commented:
Hi @liggitt ! Gentle bump for review when you get a chance. Thank you :)

cblecker committed:
Move remaining cert helper functions to client-go/util/cert
- Move public key functions to client-go/util/cert
- Move pki file helper functions to client-go/util/cert
- Standardize on certutil package alias
- Update dependencies to client-go/util/cert

cblecker commented:
@liggitt Rebased again and tests are green. Do you have cycles to review this? Thank you!

liggitt commented on staging/src/k8s.io/client-go/util/cert/pem.go:
```diff
@@ -106,6 +108,48 @@ func ParsePrivateKeyPEM(keyData []byte) (interface{}, error) {
+
+		// get PEM bytes for just this block
+		blockData := pem.EncodeToMemory(block)
+		if privateKey, err := jwt.ParseRSAPrivateKeyFromPEM(blockData); err == nil {
```
I'd rather make the x509 parse calls in directly than add a godep on this to so many kube components... with private helpers like these, we wouldn't need anything other than the stdlib:

```go
func ParsePublicKeysPEM(keyData []byte) ([]interface{}, error) {
	var block *pem.Block
	keys := []interface{}{}
	for {
		// read the next block
		block, keyData = pem.Decode(keyData)
		if block == nil {
			break
		}

		if publicKey, err := parsePublicKey(block.Bytes); err == nil {
			keys = append(keys, publicKey)
			continue
		}
		if privateRSAKey, err := parsePrivateRSAKey(block.Bytes); err == nil {
			keys = append(keys, &privateRSAKey.PublicKey)
			continue
		}
		if privateECDSAKey, err := parsePrivateECDSAKey(block.Bytes); err == nil {
			keys = append(keys, &privateECDSAKey.PublicKey)
			continue
		}

		// tolerate non-key PEM blocks for backwards compatibility
	}

	if len(keys) == 0 {
		return nil, fmt.Errorf("data does not contain any valid RSA or ECDSA public keys")
	}
	return keys, nil
}

func parsePublicKey(data []byte) (interface{}, error) {
  ...
}

func parsePrivateRSAKey(data []byte) (*rsa.PrivateKey, error) {
  ...
}

func parsePrivateECDSAKey(data []byte) (*ecdsa.PrivateKey, error) {
  ...
}
```

cblecker commented:
/retest

cblecker commented:
@liggitt Changes made and tests are passing. PTAL :)

If all looks good, I'll squash before merge.

liggitt commented:
/lgtm

cblecker commented:
/assign @deads2k 
for approval

deads2k commented:
/approve
